### ISSUE TITLE
http2 connection resets not handled

### ISSUE THREAD

**Issue by mm-brn** [DevExpress/testcafe-hammerhead#2653]
### What is your Scenario?
Using hammerhead with testcafe fails randomly with ERR_HTTP2_GOAWAY_SESSION errors if host supports http2.
See https://github.com/DevExpress/testcafe/issues/6234
Before http2 support was added we did not see any network errors like this. So I suspect it is related to the new http2 integration.
 
### What is the Current behavior?
Unhandled promise rejections if ERR_HTTP2_GOAWAY_SESSION was received. http2 session is not recreated and connection reset is not handled (unlike http1, see https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/request-pipeline/connection-reset-guard.ts)
 
### What is the Expected behavior?
Establish new connection to ensure further requests are working.
 
### What is your public web site URL?

Your website URL (or attach your complete example): 

<details>
<summary>Your complete app code (or attach your test files):</summary>

 
```js
 
```
</details>

<details>
<summary>Screenshots:</summary>

```

```
</details>

### Steps to Reproduce:

Running many testcafe tests against a server which handles http1 and http2 requests.
 
### Your Environment details:

* node.js version:                     
* browser name and version: 
* platform and version:          
* other:

**Comment by VasilyStrelyaev** [comment]
Thank you for reporting this. However, we need an example to reproduce this error on our side.
 
I have tried running a basic test against `github.com`, which seems to use HTTP/2:

<img width="548" alt="image" src="https://user-images.githubusercontent.com/11459924/122722680-afab1500-d27a-11eb-8df7-9c5e5a40abf3.png">

```js
import { Selector } from 'testcafe';
fixture `Many tests`
    .page `https://github.com`;
for(let i = 0; i < 20; i++) {
    test(`Test No. ${i}`, async t => {
        await t
            .click('a[href="#home-automate"]')
            .click(Selector('a').withText('Learn more about GitHub Actions'))
            .click('a.btn-mktg');
    });
}
```
 
This test runs without errors. Could you please send a sample (or modify this test) so that the error appears?

**Comment by mm-brn** [comment]
unfortunately this happens at random times. My only suspicion is that the server ends the long living http2 connection (session) and hammerhead is still trying to reuse the old session. (see http2 rfc https://datatracker.ietf.org/doc/html/rfc7540#section-6.8 )
From my point of view hammerhead should take care of that and reestablish the connection.

@sam-yh can you add more details?

**Comment by sam-yh** [comment]
Sorry - not really. I can just confirm that it happens randomly and rarely (though often enough to be a problem). Can't provide a publicly available page to reproduce it against. But I can say, I only saw it happening within our staging environments.

**Comment by aeneasr** [comment]
We are also seeing this issue but unfortunately I can not yet provide a reproducible test case.

**Comment by aeneasr** [comment]
I was able to create a reproducible case, here you go:

```
import { Selector } from 'testcafe';
fixture `Many tests`
    .page `https://console.ory.sh`;
for(let i = 0; i < 120; i++) {
    test(`Test No. ${i}`, async t => {
        await t
            .navigateTo('https://console.ory.sh/login')
            .click('a[href="/registration"]')
            .click(Selector('h4').withText('Become an early access member.'))
            .click('a[href="/login"]')
        await t.wait(1000)
    });
}
```

I believe the problem comes from `navigateTo`

```
 Many tests
(node:62838) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
(Use `node --trace-warnings ...` to show where the warning was created)
 ✓ Test No. 0  (15s)
 ✓ Test No. 1  (15s)
 ✓ Test No. 2  (13s)
 ✓ Test No. 3  (13s)
 ✓ Test No. 4  (12s)
 ✓ Test No. 5  (13s)
 ✓ Test No. 6  (12s)
 ✓ Test No. 7  (12s)
 ✓ Test No. 8  (12s)
 ✓ Test No. 9  (14s)
 ✓ Test No. 10  (13s)
 ✓ Test No. 11  (12s)
 ✓ Test No. 12  (12s)
 ✓ Test No. 13  (12s)
 ✓ Test No. 14  (12s)
 ✓ Test No. 15  (13s)
 ✓ Test No. 16  (11s)
 ✓ Test No. 17  (11s)
 ✓ Test No. 18  (11s)
 ✓ Test No. 19  (11s)
 ✓ Test No. 20  (12s)
 ✓ Test No. 21  (11s)
 ✓ Test No. 22  (12s)
 ✓ Test No. 23  (13s)
 ✓ Test No. 24  (11s)
 ✓ Test No. 25  (12s)
 ✖ Test No. 26  (11s)

   1) Unhandled promise rejection:
      
      Error [ERR_HTTP2_GOAWAY_SESSION]: New streams cannot be created after receiving a GOAWAY
      at new NodeError (node:internal/errors:363:5)
      at ClientHttp2Session.request (node:internal/http2/core:1705:13)
      at DestinationRequest._sendRealThroughHttp2 (/Users/foobar/workspace/go/src/github.com/ory-corp/cloud/test/e2e/node_modules/testcafe-hammerhead/lib/request-pipeline/destination-request/index.js:47:32)
      at DestinationRequest._send (/Users/foobar/workspace/go/src/github.com/ory-corp/cloud/test/e2e/node_modules/testcafe-hammerhead/lib/request-pipeline/destination-request/index.js:99:18)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

      Browser: Firefox 89.0 / macOS 10.15


 Many tests Duration (3m 00s)

 94/120 failed (3m 00s)
 Total time (3m 08s)
```

* Testcafe version: 1.14.2
* Node version: v16.3.0 & v14

*testcaferc.record.json*

```
{
  "screenshots": {
    "path": "screenshots"
  },
  "videoOptions": {
    "failedOnly": false,
    "pathPattern": "${FIXTURE}/${TEST}/${DATE}-${TIME}-${FILE_INDEX}.mp4"
  },
  "videoEncodingOptions": {
    "vcodec": "libx264",
    "preset": "veryfast"
  },
  "quarantine": false,
  "browsers": "firefox:headless",
  "videoPath": "videos",
  "disableScreenshots": false,
  "stopOnFirstFail": true,
  "speed": 1,
  "selectorTimeout": 20000,
  "assertionTimeout": 10000,
  "pageLoadTimeout": 20000,
  "pageRequestTimeout": 30000,
  "retryTestPages": true,
  "disableMultipleWindows": true,
  "src": ["tests/github.ts"],
  "skipJsErrors": true,
  "reporter": "spec-time",
  "concurrency": 2,
  "hostname": "localhost"
}
```

**Comment by aeneasr** [comment]
I updated the code sample because it appears that `t.waitFor` is not causing the issue. Please also note that it's not consistently reproducing the issue so sometimes I have to run the test a couple of times. Here's another failure:

```
 ✓ Test No. 0  (14s)
 ✓ Test No. 1  (14s)
 ✓ Test No. 2  (13s)
 ✓ Test No. 3  (13s)
 ✓ Test No. 4  (13s)
 ✓ Test No. 5  (14s)
 ✓ Test No. 6  (12s)
 ✓ Test No. 7  (12s)
 ✓ Test No. 8  (12s)
 ✓ Test No. 9  (12s)
 ✓ Test No. 10  (13s)
 ✓ Test No. 11  (13s)
 ✓ Test No. 12  (12s)
 ✓ Test No. 13  (12s)
 ✓ Test No. 14  (12s)
 ✓ Test No. 15  (13s)
 ✓ Test No. 16  (11s)
 ✓ Test No. 17  (11s)
 ✓ Test No. 18  (12s)
 ✓ Test No. 19  (11s)
 ✓ Test No. 20  (16s)
 ✓ Test No. 21  (15s)
 ✓ Test No. 22  (12s)
 ✓ Test No. 23  (12s)
 ✓ Test No. 24  (13s)
 ✓ Test No. 25  (14s)
 ✖ Test No. 26  (26s)

   1) Unhandled promise rejection:
      
      Error [ERR_HTTP2_GOAWAY_SESSION]: New streams cannot be created after receiving a GOAWAY
      at new NodeError (node:internal/errors:363:5)
      at ClientHttp2Session.request (node:internal/http2/core:1705:13)
      at DestinationRequest._sendRealThroughHttp2 (/Users/foobar/workspace/go/src/github.com/ory-corp/cloud/test/e2e/node_modules/testcafe-hammerhead/lib/request-pipeline/destination-request/index.js:47:32)
      at DestinationRequest._send (/Users/foobar/workspace/go/src/github.com/ory-corp/cloud/test/e2e/node_modules/testcafe-hammerhead/lib/request-pipeline/destination-request/index.js:99:18)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

      Browser: Firefox 89.0 / macOS 10.15
```

Interestingly, the 26th test run takes 26 seconds vs the 14s of the previous one and then fails. Please also note that we are **running tests with parallelism of 2**.

**Comment by mm-brn** [comment]
@aeneasr 

the tests are still running (Test No. 176) without HTTP2_GOAWAY... I guess you ran the tests during work. Was there any deploy while running the test? Maybe a deploy is causing nginx or whatever webserver you are using to close connections.

**Comment by aeneasr** [comment]
No, we didn’t run any deploys during that time. One possibility could be location - I am in EU but the servers are in US Central. But I also noticed that it’s not 100% reproducible. Sometimes I have to run it a few times to observe the issue.

But maybe the stack trace is helping?

**Comment by aeneasr** [comment]
Looking through some sources it looks like this is an issue in hammerhead. Generally, GOAWAY should be handled gracefully by the client (see [this answer for example](https://stackoverflow.com/questions/55087292/how-to-handle-http-2-goaway-with-httpclient)) as the server is allowed to close the connection at any time for any, or even no, reason at all. Since `GOAWAY` is not handled by hammerhead, this results in an unhandled exception.

I'm unfortunately not a node wizard, but it looks to me as if we would need to handle goaway somewhere around here:

https://github.com/DevExpress/testcafe-hammerhead/blob/ea5d6db645794caf6d16684ecf07f131c5bcab7e/src/request-pipeline/destination-request/index.ts#L62

Probably [something like](https://nodejs.org/api/http2.html#http2_event_goaway):

```
stream.on('goaway', () => {
  // Somehow create a new stream or downgrade to http 1.0
})
```

**Comment by aeneasr** [comment]
Here's some more information, looks like gateways such as AWS ELB / ALB can also shutdown HTTP 2.0 connections without reason: https://stackoverflow.com/questions/41592929/aws-alb-http-2-and-goaway

I looked a bit more through the source code and it appears that ERRCONNRESET and similar are handled, so I think the logic to retry on connection resets is there, it's just that `GOAWAY` is not properly respected in that regard!

**Comment by aeneasr** [comment]
I believe we have a workaround which is downgrading `testcafe-hammerhead` to `23.0.0` as that has no HTTP2 support.

**Comment by felis2803** [comment]
A similar error occurs when running a test like this:
```js
import { Selector, Role } from 'testcafe';

fixture `xaf-demos`
    .page `https://demos.devexpress.com/XAF`;

const role = Role('https://demos.devexpress.com/XAF/MainDemo', async t => {
    await t
        .typeText('.dxeEditArea_XafTheme.dxeEditAreaSys', 'John', { replace: true })
        .click(Selector('span').withText('Log In'));
});

test('Role test', async t => {
    await t.useRole(role);
});
```
Exception:
```
Error in Role initializer
TestCafe encountered an error while executing this test. 

Please report this problem to the Support Center: http://www.devexpress.com/ask.

Error [ERR_HTTP2_ERROR]: The user callback function failed
at Http2Session.onSessionInternalError (internal/http2/core.js:751:26)
at Http2Session.callbackTrampoline (internal/async_hooks.js:131:14)

Browser: Chrome 91.0.4472.124 / Windows 10
```

**Comment by aeneasr** [comment]
I can confirm that downgrading to 23.0.0 resolves all HTTP2-related errors!

**Comment by felis2803** [comment]
@aeneasr, thanks for the advice. However, this method will not work for my project.

**Comment by aeneasr** [comment]
The downgrade no longer works with testcafe > 1.15.0 unfortunately.

**Comment by aeneasr** [comment]
Awesome, thank you for the fix @LavrovArtem !

**Comment by aeneasr** [comment]
If a release for this would be possible, I would highly appreciate it! This has been blocking us for quite some time :) If not, we'll just wait until someone with write access has time :)

**Comment by aleks-pro** [comment]
We will include this fix in the next minor release - v1.15.2. In the meantime, you can use v1.15.2-rc.1, which contains the fix.

**Comment by sam-yh** [comment]
> 
> 
> We will include this fix in the next minor release - v1.15.2. In the meantime, you can use v1.15.2-rc.1, which contains the fix.

I can confirm that with 1.15.2-rc.1 the issue does not occur anymore! Thanks! Looking forward to 1.15.2
